### DISCUSSION TITLE
Error while creating CSRF token: Cannot read properties of undefined (reading 'digest')

### DISCUSSION THREAD

**Discussion by nowaythatworked** [nextauthjs/next-auth#6418] [Help]
### Environment

System:
    OS: macOS 12.2.1
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Memory: 6.82 GB / 32.00 GB
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 16.18.0 - ~/.nvm/versions/node/v16.18.0/bin/node
    Yarn: 1.22.19 - ~/.yarn/bin/yarn
    npm: 8.19.2 - ~/.nvm/versions/node/v16.18.0/bin/npm
  Browsers:
    Chrome: 109.0.5414.87
    Safari: 15.3

### Reproduction URL

-

### Describe the issue

I wanted to use @auth/core with astro. I put together something similar to the solid-js package. Unfortunately I can´t get up running. While logging in I first request a CSRF token from '/api/auth/csrf', during this I get the error 
`TypeError: Cannot read properties of undefined (reading 'digest')
cms:dev:     at createHash (file:///Users/<censored>/projects/lite/litecms/node_modules/@auth/core/lib/web.js:89:38)
cms:dev:     at createCSRFToken (file:///Users/<censored>/projects/lite/litecms/node_modules/@auth/core/lib/csrf-token.js:19:45)
cms:dev:     at init (file:///Users/<censored>/projects/lite/litecms/node_modules/@auth/core/lib/init.js:76:73)
cms:dev:     at AuthInternal (file:///Users/<censored>/projects/lite/litecms/node_modules/@auth/core/lib/index.js:8:40)
cms:dev:     at Proxy.Auth (file:///Users/<censored>/projects/lite/litecms/node_modules/@auth/core/index.js:95:36)
cms:dev:     at processTicksAndRejections (node:internal/process/task_queues:96:5)
cms:dev:     at async eval (/Users/<censored>/projects/lite/litecms/packages/auth/src/server.ts:48:12)
cms:dev:     at async Module.get (/Users/<censored>/projects/lite/litecms/packages/auth/src/server.ts:34:33)`

I replaced the used crypto lib with webcrypto (import { webcrypto } from 'crypto'). This works for csrf creation but then I get an error when requesting the signIn endpoint. 

I also deployed it to vercel, where I get following error:
`ERROR	Unhandled Promise Rejection 	{"errorType":"Runtime.UnhandledPromiseRejection","errorMessage":"ReferenceError: crypto is not defined"`

### How to reproduce

1. Setup a new Astro Project with the latest version. 
2. Use exported Auth function from @auth/core for auth endpoints.
3. In frontmatter request /api/auth/csrf.

### Expected behavior

Getting a CSRF token on request and being able to use it for login.

**Comment by tajkirkpatrick** [comment]
Alongside the AstroJS community (within the AstroJS Discord), I am also trying to integrate AuthJs and Astro. I created #6373 , which got turned into a discussion, and ultimately PR #6379 would fix both of our issues. Although losing CSRF might not be necessary anyway. (Unless you know how to implement it with Astro.cookies, CSRF is important to keep if possible.) 

That being said a mod, shout-out Otterlord, within the discord community forked a successful repo and example. You'll continue to see this crypto.subtle == undefined error UNLESS your node version > 17.4.0

Mr. Otterlord says there might be a case to support a next-auth PR that allows for the node version backwards compatibility to around v15 or v16. More info and support in the Discord, for a full catch up.  

See:
https://nodejs.org/api/crypto.html#cryptosubtle

**Comment by nowaythatworked** [comment]
@tajkirkpatrick Thanks for your reply! Can you explain how it is connected to Astro cookies?
I would really like to find a solution with CSRF tokens, I will join the discord and contact you there :)

**Comment by tajkirkpatrick** [comment]
Yeah no problem, as you have found the crypto web api is undefined in a server runtime context. 
I mentioned Astro.cookies because let us assume you create a hypothetical AstroJS wrapper for AuthJS that defaults to using the newer skipCSRFchecks flag in PR #6379. (because this would get you up and running the quickest) 

To make sure the Astro site is CSRF protected, YOU the framework author would have to create the CSRF implementation inside this hypothetical @authjs/astro package. The AstroJS framework offers the Astro.cookies.set(), Astro.cookies.get(), Astro.cookies.etc..... utilities to manipulate cookie behavior, (you could also use HTTP headers but I'm no expert in getting headers setup PER request, in AstroJS... but I digress)

You could theoretically protect pages on your Astro site from being accessible by some sort of control flow similar to the one below.

```astro
---
const authenticated = Astro.cookies.get('authjs-astro-cookie')

if (!authenticated) Astro.redirect('/')
---
<html>...</html>
```
That is why I was mentioning an Astro.cookies implementation of CSRF protection would be needed if you created your own package. Lastly though, you'll find in the Astro Discord (support channel) that this error with createHash you're encountering solves itself when you use a Node version > 17.4.0. That being said this Discussion #6417 runs into our same error even though they're on Node v18. (Albeit not using AstroJS though).

Did that explanation help @nowaythatworked ?

**Comment by tajkirkpatrick** [comment]
Edit: AstroJS offers header manipulation at the page level also, using Astro.request.headers , so is possible to implement our very own CSRF protection, but I think making AuthJS Node v15 compatible solves more compatibility issues with the Crypto API in the long run.

https://docs.astro.build/en/guides/server-side-rendering/#astrorequestheaders

**Comment by nowaythatworked** [comment]
Yeah, unfortunately there is nothing like a middleware in astro :/ 
I will take a look tomorrow, maybe it´s possible to have an intermediate solution.

thanks again, you helped a lot

**Comment by balazsorban44** [comment]
Converted to discussion as it's more like discussing Astro support. 

For the original issue, see my answer here:
https://github.com/nextauthjs/next-auth/issues/6417#issuecomment-1384660656
